packages/common/src/subchain/EdenSubchain.ts

```diff
@@ -0,0 +1,215 @@
+        if (this.corrupt) throw new Error("wasm state is corrupt");
+        return new Uint8Array(this.memory!.buffer, pos, len);
+    }
+
+    decodeStr(pos?: number, len?: number) {
+        return new TextDecoder().decode(this.uint8Array(pos, len));
+    }
+
+    resultAsUint8Array() {
+        return this.uint8Array(
+            this.exports.getResult(),
+            this.exports.getResultSize()
+        );
+    }
+
+    resultAsString() {
+        return this.decodeStr(
+            this.exports.getResult(),
+            this.exports.getResultSize()
+        );
+    }
+
+    imports = {
+        clchain: {
+            abort_message: (pos: number, len: number) => {
+                throw new Error("abort: " + this.decodeStr(pos, len));
+            },
+            console: (pos: number, len: number) => {
+                const s = this.consoleBuf + this.decodeStr(pos, len);
+                const l = s.split("\n");
+                for (let i = 0; i < l.length - 1; ++i) console.log(l[i]);
+                this.consoleBuf = l[l.length - 1];
+            },
+        },
+    };
+
+    async instantiate(module: BufferSource) {
+        const result = await WebAssembly.instantiate(module, this.imports);
+        this.module = result.module;
+        this.instance = result.instance;
+        this.exports = result.instance.exports;
+        this.memory = this.exports.memory;
+        return this;
+    }
+
+    async instantiateStreaming(response: Response | PromiseLike<Response>) {
+        const result = await WebAssembly.instantiateStreaming(
+            response,
+            this.imports
+        );
+        this.module = result.module;
+        this.instance = result.instance;
+        this.exports = result.instance.exports;
+        this.memory = this.exports.memory;
+        return this;
+    }
+
+    protect<T>(f: () => T) {
+        if (this.corrupt) throw new Error("wasm state is corrupt");
+        try {
+            return f();
+        } catch (e) {
+            this.corrupt = true;
+            throw e;
+        }
+    }
+
+    initializeMemory(
+        edenAccount: string,
+        tokenAccount: string,
+        atomicAccount: string,
+        atomicmarketAccount: string
+    ) {
+        if (this.initialized)
+            throw new Error("wasm memory is already initialized");
+        const buf = new Serialize.SerialBuffer();
+        buf.pushName(edenAccount);
+        buf.pushName(tokenAccount);
+        buf.pushName(atomicAccount);
+        buf.pushName(atomicmarketAccount);
+        const edenAccountLow = buf.getUint32();
+        const edenAccountHigh = buf.getUint32();
+        const tokenAccountLow = buf.getUint32();
+        const tokenAccountHigh = buf.getUint32();
+        const atomicAccountLow = buf.getUint32();
+        const atomicAccountHigh = buf.getUint32();
+        const atomicmarketAccountLow = buf.getUint32();
+        const atomicmarketAccountHigh = buf.getUint32();
```

Comment:
Since the `initialize` function initializes memory, it's not possible for it to accept an array as input, since that requires calling `allocateMemory` first, which doesn't work on uninitialized memory. This would require splitting `initialize` into 2 functions.